### PR TITLE
CTD - Python implementation

### PR DESCRIPTION
**Pull request by NevenaaNikolic**
Python implementation of the CTD algorithm. Aligned with the R one.

### CODE REVIEW

File: CTD.py [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+from scipy.stats import norm
+from sklearn import covariance
+import json
+from multiprocessing import Pool
+from functools import partial
+import pandas as pd
+from Python import data, graph, mle
+import logging
+import numpy as np
+import os
```
Comment (vladimirkovacevic) [inline CTD.py]: Add imports from the standard Python library first, then empty line then other, from pip-installed libraries

File: CTD.py [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+            logging.debug('Node "{}" not in graph. Exiting program.'.format(node))
+            exit(1)
+
+    # Walk through all the nodes in S module
+    logging.debug('Get the single-node encoding node ranks starting from each node.')
+
+    # Dictionary ranks contains encodings for each node in S_perturbed_nodes
+    ranks = {}
+
+    if argv.num_processes == 1:
```
Comment (vladimirkovacevic) [inline CTD.py:125]: Add comment that this if is added to avoid overhead

File: CTD.py [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+            ranks[tup[1]] = tup[0]
+
+    # Convert to bitstring
+    # Get the bitstring associated with the disease module's metabolites
+    pt_bs_by_k = mle.get_pt_bs_by_k(S=S_perturbed_nodes, ranks=ranks)
+
+    # Get encoding length of minimum length code word.
+    # experimental_df is dataframe with diseases (and surrogates)
+    # and z-values for each metabolite
+    # TODO: If graph and disease module are given do we still need experimental_df?
```
Comment (vladimirkovacevic) [inline CTD.py]: we don't. remove todo

File: CTD.py [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+    # You can interpret the probability assigned to this metabolite set by
+    # comparing it to a null encoding algorithm, which uses fixed-length codes
+    # for all metabolites in the set. The "d.score" is the difference in bitlength
+    # Significance theorem, we can estimate the upper bounds on a p-value by 2^-d.score.
+
+    p_value_F = 2.0 ** (-1 * F_info['d.score'])
+    # All metabolites in the bitstring
+    logging.debug(f'All metabolites in the bitstring: {[d[0] for d in pt_bs_by_k[ind_F]]}')
+
+    # Just the F metabolites that are in S_arg that were were "found"
```
Comment (vladimirkovacevic) [inline CTD.py]: S_arg?

File: CTD.py [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+p.add_argument("--out_graph_name", help="Name of the output graph adjacency CSV file.")
+p.add_argument("--num_processes", help="Number of worker processes to use for parallelisation. Default is to use the "
+                                       "number returned by os.cpu_count().", default=cpu_count, type=int)
+p.add_argument("-v", "--verbose", action="count", help="Enable verbose logging.")
+
+if __name__ == '__main__':
+
+    argv = p.parse_args()
+
+    logging.getLogger().setLevel(logging.DEBUG)
```
Comment (vladimirkovacevic) [inline CTD.py]: Control logging level by --verbose parameter

File: CTD.r [hunk 6]
```diff
@@ -193,3 +193,8 @@ if (is.na(argv$output_name)){
   outfname = argv$output_name
 }
 write(res_json, outfname)
+
+res_ranks = toJSON(ranks)
+outrname = str_replace(outfname, '.json', '_ranks.json')
+write(res_ranks, outrname)
```
Comment (vladimirkovacevic) [inline CTD.r:199]: I would output this file if verbose >=2.

File: Python/mle.py [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+
+    res = []
+
+    for k in range(len(bs)):
+        row = {}
+        opt_bs = bs[k]
+        mets_k = [d[0] for d in opt_bs if d[1] == 1]
+        found = sum([d[1] for d in opt_bs])
+        not_found = k - found + 1
+        e = (not_found + 1) * np.log2(len(G)) + len(opt_bs) - 1
```
Comment (vladimirkovacevic) [inline Python/mle.py]: add few comments inside function

File: R/data.surrogateProfiles.r [hunk 8]
```diff
@@ -75,5 +75,7 @@ data.surrogateProfiles = function(data, std=1, ref_data) {
         data_mx_surr=data_mx_surr[-which(var.met == 0),]}
     rownames(data_mx_surr) = tolower(rownames(data_mx_surr))
     data_mx_surr = apply(data_mx_surr, c(1, 2), as.numeric)
-    return(data_mx_surr)
+    
+    # add as.data.frame for calling from Python
+    return(as.data.frame(data_mx_surr))
```
Comment (vladimirkovacevic) [inline R/data.surrogateProfiles.r]: Will this brake R execution?

File: README.md [hunk 9]
```diff
@@ -12,34 +12,39 @@ CTD can receive as input experimental (disease) and control datasets with z-scor
 Another option is to provide a weighted graph (adjacency matrix) and a list of nodes of interest.
```
Comment (vladimirkovacevic) [inline README.md:12]: Resolve conflicts with latest master.

File: R/singleNode.getNodeRanksN.r [hunk 10]
```diff
@@ -59,6 +59,10 @@
 singleNode.getNodeRanksN = function(n,G,p1,thresholdDiff,adj_mat,S=NULL,
                                     num.misses=NULL,verbose=FALSE,out_dir="",
                                     useLabels=FALSE,coords=NULL) {
+    
+    rownames(adj_mat) = names(G)
```
Comment (vladimirkovacevic) [inline R/singleNode.getNodeRanksN.r]: do we need this?

File: R_vs_Python_comparison.py [hunk 11]
```diff
@@ -0,0 +1,117 @@
+import argparse
```
Comment (vladimirkovacevic) [inline R_vs_Python_comparison.py:1]: Add imports from the standard Python library first, then empty line then other, from pip-installed libraries

**Review by vladimirkovacevic — CHANGES_REQUESTED** [review]

File: R/data.surrogateProfiles.r (hunk 8 above)
Comment (NevenaaNikolic) [reply R/data.surrogateProfiles.r]: It is possible, please ignore all changes made to the R code (except the ranks saving in the main CTD script).

File: R/singleNode.getNodeRanksN.r (hunk 10 above)
Comment (NevenaaNikolic) [reply R/singleNode.getNodeRanksN.r]: No, we don't - this was added so the Python-R code can work as NumPy matrices do not have row/column names.

File: README.md (hunk 9 above)
Comment (NevenaaNikolic) [reply README.md:12]: Resolved.

**Comment by NevenaaNikolic** [comment]
All requested changes reviewed & implemented.

**Review by vladimirkovacevic — APPROVED** [review]

**Pull request merged**